src/Development/Shake/Command.hs

```diff
@@ -561,6 +561,15 @@ type a :-> t = a
 cmd :: CmdArguments args => args :-> Action r
 cmd = cmdArguments []
 
+class Unit a
+instance {-# OVERLAPPING #-} Unit b => Unit (a -> b)
+instance {-# OVERLAPPABLE #-} a ~ () => Unit (m a)
```

Comment:
The `OVERLAPPING` and `OVERLAPPABLE` are not available until GHC 7.10, and Shake still supports back to GHC 7.4. I think you can tweak it to be a language pragma on those versions - may require CPP though.